### ISSUE TITLE
Custom objective function documentation clarifications

### ISSUE THREAD

**Issue by mstewart141** [catboost/catboost#521]
Problem: Request for documentation clarification
catboost version: 0.10.3
Operating System: darwin
CPU and GPU

Hi all, first off, thanks for the great library!

Here is the documentation page on which I seek clarification (my questions are below):
https://tech.yandex.com/catboost/doc/dg/concepts/python-usages-examples-docpage/#custom-objective-function

The heart of which is:
```
class LoglossObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        # approxes, targets, weights are indexed containers of floats (containers with only __len__ and __getitem__ defined).
        # weights parameter can be None.
        # Returns list of pairs (der1, der2)
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        exponents = []
        for index in xrange(len(approxes)):
            exponents.append(math.exp(approxes[index]))

        result = []
        for index in xrange(len(targets)):
            p = exponents[index] / (1 + exponents[index])
            der1 = (1 - p) if targets[index] > 0.0 else -p
            der2 = -p * (1 - p)

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))

        return result
```

1) I am curious why pure python is used here and not numpy? Is this required or is allowable to use numpy? 

2) If I would like the loss to work on GPU as well as CPU is that possible? If it is possible, must it use pure python, or is numpy allowable?

3) Are data (`approxes`, `target`) delivered in "batch" aka the whole dataset, or in mini-batch? Are the data shuffled or will they always appear in the order in which the data (here a large `np.array`) are passed to the `fit` method? This is very important to know for metrics that can not be calculated purely at the row level: I would like to aggregate rows in my loss function, and this is very easy to do as long as the data appear in a consistent ordering in "batch". **TL;DR** If I feed catboost `.fit` a million row numpy array, will i get approxes and target both of length 1million in the same order as the input data as the input to this `calc_ders_range` (if not what is the behaviour)?

4) Here the `calc_ders_range` takes `approxes` and `targets`: is it possible to access the rows of data associated with these predictions as well? I.e., if I would like to know the value of a feature for the given row in order to use that in the loss function, is there any way to accomplish this?

Thank you again!

**Comment by annaveronika** [comment]
- Writing loss function in python is not recommended, because it will work very slow. It's much better to do it in C++. You can look in our tutorials for examples.
- It will not work for GPU training.
- approxes and target are fully stored in memory
- it is not possible to access rows of data inside this particular functions

**Comment by annaveronika** [comment]
@mstewart141 Do you need any additional information?

**Comment by mstewart141** [comment]
no; thank you!

**Comment by Oleg67** [comment]
Hi 
How I can implement my custom loss function for Regrasion task on Python?

**Comment by jface1616** [comment]
Hi quick clarification on above. I understand custom loss functions in python are not supported on GPU, but are they for C++ custom loss functions? If not, are there any plans to support this in the near future?

**Comment by Oleg67** [comment]
Hi
Yes the function in Python and it is slow. I'm going to re-write it by C.

пн, 23 нояб. 2020 г. в 22:55, jface1616 <notifications@github.com>:

**Comment by taojiaming19961115** [comment]
Hi,How can I implement my custom loss function for Regrasion task on Python?
